### PR TITLE
Fixed #27021 -- Allowed lookup expressions in annotations, aggregations, and QuerySet.filter().

### PR DESCRIPTION
**Pull request by Ian-Foote**
https://code.djangoproject.com/ticket/27021

### CODE REVIEW

File: django/db/models/lookups.py [hunk 1]
```diff
@@ -293,6 +316,11 @@ def as_postgresql(self, compiler, connection):
 class Exact(FieldGetDbPrepValueMixin, BuiltinLookup):
     lookup_name = 'exact'
 
+    def __init__(self, *args, **kwargs):
+        from django.db.models import BooleanField
+        self.output_field = BooleanField()
```
Comment (hannseman) [inline django/db/models/lookups.py:321]: Are these local imports really needed? Looks like this module imports `django.db.models.fields` just fine?

File: django/db/models/lookups.py [hunk 2]
```diff
@@ -163,6 +174,18 @@ def __eq__(self, other):
...
+        c.lhs = self.lhs.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        c.rhs = self.rhs.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return c
+
+    def select_format(self, compiler, sql, params):
+        return sql, params
+
+    def get_db_converters(self, connection):
+        return []
+
```
Comment (charettes) [inline django/db/models/lookups.py:188]: Have you tried subclassing `Expression` instead of redefining all of these methods? Looks like a lot the `Lookup` boilerplate could go away with

```python
class Lookup(Expression):
    ...

    def __init__(self, lhs, rhs):
        self.lhs, self.rhs = lhs, rhs
        super().__init__(lhs, rhs)
        ...

    @cached_property
    def output_field(self):
        return BooleanField()

    ...
```

**Review by charettes — COMMENTED** [review]
Thanks for the patch Ian! [We've had recent questions about this exact usage cases](https://forum.djangoproject.com/t/why-no-query-expression-for-equality-checks/7103).

I think we should defer the decision around `__gt__` and friends and focus on making `Lookup` implement as much of the _expression_ API as possible. That will not only pave the way for annotations but also direct usage in `filter(...)` and friends which can be useful under certain circumstances e.g. https://groups.google.com/g/django-developers/c/iiOnxAn3vy4

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -146,6 +146,22 @@ def __ror__(self, other):
...
+        return GreaterThanOrEqual(self, other)
+
+    def __lt__(self, other):
+        from django.db.models.lookups import LessThan
+        return LessThan(self, other)
+
+    def __le__(self, other):
+        from django.db.models.lookups import LessThanOrEqual
+        return LessThanOrEqual(self, other)
+
```
[the remark this answers is not in the document]
Comment (Ian-Foote) [reply django/db/models/expressions.py:164]: Yeah, the difficulty with `__eq__` is why I have this bit in a second commit right now.

I'd love to be able to break backward compatibility here to allow `==` and `!=`, but I think that will require discussion on the mailing list.

File: django/db/models/lookups.py (hunk 2 above)
Comment (Ian-Foote) [reply django/db/models/lookups.py:188]: I will try this today.

File: django/db/models/expressions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ian-Foote) [reply django/db/models/expressions.py]: The second commit I refer to here has been rebased away now.

File: django/db/models/lookups.py [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
+    BooleanField, CharField, DateTimeField, Field, IntegerField, UUIDField,
 )
 from django.db.models.query_utils import RegisterLookupMixin
 from django.utils.datastructures import OrderedSet
 from django.utils.functional import cached_property
 from django.utils.hashable import make_hashable
 
 
-class Lookup:
+class Lookup(CombinedExpression):
```
Comment (felixxm) [inline django/db/models/lookups.py:17]: Why the `CombinedExpression` and not `Expression`? IMO it's misleading, I know that `CombinedExpression` has the concept of right-hand and left-hand sides but for other purposes.

**Review by felixxm — COMMENTED** [review]
@Ian-Foote Great job, thanks :rocket: 

Release notes are missing. I would also

- add examples in the [Query Expressions -> Some examples](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#some-examples),
- add a short paragraph in the [Built-in Expressions](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#built-in-expressions) or [Lookup](https://docs.djangoproject.com/en/3.2/ref/models/lookups/) docs :thinking: 


Can we add extra tests? 
- combined lookups, e.g. `Exact(F('integer'), 42) | GreaterThan(F('integer'), 42)`,
- using lookups directly in the `QuerySet.filter()`, e.g. `.filter(GreaterThan(F('integer'), 42))`,
- using lookups in the `Queryset.alias()`.

File: django/db/models/lookups.py (hunk 4 above)
Comment (felixxm) [reply django/db/models/lookups.py:17]: > both `CombinedExpression` and `Lookup` combine a left and a right expression.

Not always, many lookups are complicated expressions or even function calls. The only common factor for me is that both have two arguments. Also `CombinedExpression` has a lot of unnecessary logic, e.g. `SQLiteNumericMixin`. We should probably compare `Lookup` subclassing `Expression` vs. `CombinedExpression` :thinking:

File: django/db/models/lookups.py [hunk 5]
```diff
@@ -74,10 +75,21 @@ def set_source_expressions(self, new_exprs):
...
+        if hasattr(self.lhs, 'output_field'):
+            if hasattr(self.lhs.output_field, 'get_prep_value'):
+                return self.lhs.output_field.get_prep_value(self.rhs)
+        elif self.rhs_is_direct_value():
+            return Value(self.rhs)
         return self.rhs
 
+    def get_prep_lhs(self):
+        from django.db.models.fields.related_lookups import MultiColSource
+        if hasattr(self.lhs, 'resolve_expression') or isinstance(self.lhs, MultiColSource):
```
Comment (charettes) [inline django/db/models/lookups.py:89]: I wonder if this could be addressed by adding a `MultiColSource` method that returns `self`

**Review by charettes — COMMENTED** [review]
Thanks for the adjustments Ian!

**Comment by charettes** [comment]
Note that the making `Lookup(Expression)` and implementing `.resolve_expression` will allow some simplification in `Where.resolve_expression`

https://github.com/django/django/blob/ed3af3ff4b3cfb72de598f1b39a1028ba3826efb/django/db/models/sql/where.py#L189-L209

**Comment by Ian-Foote** [comment]
@charettes I'm not clear what change you're suggesting in `Where.resolve_expression`.

**Comment by charettes** [comment]
@Ian-Foote it can be handled in a following the PR. The idea is that `Where.resolve_expression` does a lot of duck-type digging where it could possibly simply call `resolve_expression` on all it's children now.

e.g. (still untested)

```python
def resolve_expression(self, *args, **kwargs): 
    clone = self.clone()
    clone.children = [
        child.resolve_expression(*args, **kwargs) for child in clone.children
    ]
    clone.resolved = True
    return clone
```

**Comment by Ian-Foote** [comment]
@charettes I just tested that and got a bunch of errors. I spent a little time trying to debug, but decided to leave it for now.

**Comment by charettes** [comment]
@Ian-Foote sounds good, thanks for having a look!

**Comment by Ian-Foote** [comment]
@felixxm Good call on the combined expressions test - that needed some more tweaks to `WhereNode`.

File: docs/ref/models/lookups.txt [hunk 6]
```diff
@@ -177,16 +177,21 @@ following methods:
     comparison between ``lhs`` and ``rhs`` such as ``lhs in rhs`` or
     ``lhs > rhs``.
 
-    The notation to use a lookup in an expression is
+    The the primary notation to use a lookup in an expression is
     ``<lhs>__<lookup_name>=<rhs>``.
 
-    This class acts as a query expression, but, since it has ``=<rhs>`` on its
-    construction, lookups must always be the end of a lookup expression.
+    .. versionchanged:: 4.0
```
Comment (felixxm) [inline docs/ref/models/lookups.txt:183]: Blank line is missing.

File: docs/ref/models/lookups.txt [hunk 7]
```diff
@@ -177,16 +177,21 @@ following methods:
     comparison between ``lhs`` and ``rhs`` such as ``lhs in rhs`` or
     ``lhs > rhs``.
 
-    The notation to use a lookup in an expression is
+    The the primary notation to use a lookup in an expression is
```
Comment (felixxm) [inline docs/ref/models/lookups.txt:180]:
```suggestion
    The primary notation to use a lookup in an expression is
```

**Review by felixxm — COMMENTED** [review]
@Ian-Foote Thanks for updates :+1: We're close :rocket:. I pushed edits to fix failures on Oracle.

I would add a few more tests:
- using combined lookups directly in `QuerySet.filter()`,
- using annotation of combined lookups in `QuerySet.filter()`,
- using combined lookups/lookups in `QuerySet.aggregate()`.

File: django/db/models/fields/related_lookups.py [hunk 8]
```diff
@@ -22,6 +22,9 @@ def relabeled_clone(self, relabels):
     def get_lookup(self, lookup):
         return self.output_field.get_lookup(lookup)
 
+    def resolve_expression(self, *args, **kwargs):
+        return self
+
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/models/fields/related_lookups.py]: OK, we need it :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: docs/ref/models/lookups.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply docs/ref/models/lookups.txt:180]: https://github.com/django/django/pull/14494#discussion_r657683574 :wink:

[the remark this answers is not in the document]
Comment (pauloxnet) [reply docs/ref/models/lookups.txt:180]: Sorry, my old comment was pending.

File: django/db/models/fields/related_lookups.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ian-Foote) [reply django/db/models/fields/related_lookups.py:27]: I definitely had a failure - I'll double check.

[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/models/fields/related_lookups.py:27]: Yes, we need it, sorry for the noise https://github.com/django/django/pull/14494#discussion_r657703518.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by Ian-Foote** [comment]
@felixxm I added those extra tests and fixed another edge-case. I'm not sure quite how to get the docs to build correctly.

File: docs/ref/models/lookups.txt [hunk 9]
```diff
@@ -177,16 +177,22 @@ following methods:
...
 
-    The notation to use a lookup in an expression is
+    The primary notation to use a lookup in an expression is
     ``<lhs>__<lookup_name>=<rhs>``.
 
-    This class acts as a query expression, but, since it has ``=<rhs>`` on its
-    construction, lookups must always be the end of a lookup expression.
+    .. versionchanged:: 4.0
+
+        A lookup can also be used directly as an expression:
```
Comment (hannseman) [inline docs/ref/models/lookups.txt:185]:
```suggestion
        A lookup can also be used directly as an expression::
```

Think this might fix the issue with the docs not building.

Comment (Ian-Foote) [reply docs/ref/models/lookups.txt]: Doesn't look like it :disappointed:

File: tests/expressions/tests.py [hunk 10]
```diff
@@ -2045,3 +2048,124 @@ def test_hash(self):
             hash(OrderBy(F('field'), nulls_last=True)),
             hash(OrderBy(F('field'), nulls_last=False)),
         )
+
+
+class ComparisonExpressionTests(TestCase):
```
Comment (felixxm) [inline tests/expressions/tests.py:2053]: `tests/expressions` already contains a mix of tests. I would move them to the `tests/lookup/tests.py`

File: tests/expressions/tests.py [hunk 11]
```diff
@@ -2045,3 +2048,124 @@ def test_hash(self):
...
+
+    def test_alias(self):
+        numbers = Number.objects.alias(greater=GreaterThan(F('integer'), 10))
+        numbers = numbers.filter(greater=True)
+        self.assertSequenceEqual(numbers, [self.n])
+
+    def test_aggregate_combined_lookup(self):
+        expression = Cast(GreaterThan(F('integer'), 0), IntegerField())
+        positive = Number.objects.aggregate(positive=Sum(expression))
+        self.assertEqual(positive['positive'], 1)
```
Comment (felixxm) [inline tests/expressions/tests.py:2171]: This will not work on Oracle. I added `@skipUnlessDBFeature`.

File: django/db/models/lookups.py [hunk 12]
```diff
@@ -163,6 +150,20 @@ def __eq__(self, other):
     def __hash__(self):
         return hash(make_hashable(self.identity))
 
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        c = copy(self)
```
Comment (felixxm) [inline django/db/models/lookups.py:154]:
```suggestion
        c = self.copy()
```

File: django/db/models/lookups.py [hunk 13]
```diff
@@ -140,16 +135,8 @@ def as_oracle(self, compiler, connection):
...
-    def contains_aggregate(self):
-        return self.lhs.contains_aggregate or getattr(self.rhs, 'contains_aggregate', False)
-
-    @cached_property
-    def contains_over_clause(self):
-        return self.lhs.contains_over_clause or getattr(self.rhs, 'contains_over_clause', False)
-
-    @property
-    def is_summary(self):
-        return self.lhs.is_summary or getattr(self.rhs, 'is_summary', False)
```
Comment (felixxm) [inline django/db/models/lookups.py:152]: I think we should keep this property or add `c.is_summary = summarize` to the `resolve_expression()`.

File: django/db/models/lookups.py [hunk 14]
```diff
@@ -74,10 +75,20 @@ def set_source_expressions(self, new_exprs):
     def get_prep_lookup(self):
         if hasattr(self.rhs, 'resolve_expression'):
             return self.rhs
-        if self.prepare_rhs and hasattr(self.lhs.output_field, 'get_prep_value'):
-            return self.lhs.output_field.get_prep_value(self.rhs)
+        if not self.prepare_rhs:
+            return self.rhs
```
Comment (felixxm) [inline django/db/models/lookups.py:79]: We could optimize this a bit:
```python
        if not self.prepare_rhs or hasattr(self.rhs, 'resolve_expression'):
            return self.rhs
        if hasattr(self.lhs, 'output_field'):
            ...
```

**Review by felixxm — COMMENTED** [review]
@Ian-Foote Thanks for updates :+1: I skipped one test on Oracle, pushed small edits to docs, and left comments. Can we add one more test :wink:? i.e. using lookups directly in `Case()`.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/oracle/operations.py [hunk 15]
```diff
@@ -202,7 +202,7 @@ def get_db_converters(self, expression):
         # Oracle stores empty strings as null. If the field accepts the empty
         # string, undo this to adhere to the Django convention of using
         # the empty string instead of null.
-        if expression.field.empty_strings_allowed:
+        if hasattr(expression, 'field') and expression.field.empty_strings_allowed:
```
Comment (charettes) [inline django/db/backends/oracle/operations.py:205]: Wonder if we should check for `output_field` instead?

```suggestion
        if expression.output_field.empty_strings_allowed:
```

**Review by charettes — APPROVED** [review]

**Comment by charettes** [comment]
@felixxm I didn't look into details of the Oracle failure but lmk if you hit a wall.

**Comment by felixxm** [comment]
> @felixxm I didn't look into details of the Oracle failure but lmk if you hit a wall.

@Ian-Foote `test_annotation_aggregate_with_m2o` fails because `book_contact_set__isnull` is unnecessarily added to the `GROUP BY` clause:
```sql
SELECT
    "ANNOTATIONS_AUTHOR"."NAME",
    CASE WHEN "ANNOTATIONS_BOOK"."ID" IS NULL THEN 0 ELSE MAX(T4."PAGES") END AS "MAX_PAGES"
FROM
    "ANNOTATIONS_AUTHOR"
    LEFT OUTER JOIN "ANNOTATIONS_BOOK" ON ("ANNOTATIONS_AUTHOR"."ID" = "ANNOTATIONS_BOOK"."CONTACT_ID")
    LEFT OUTER JOIN "ANNOTATIONS_BOOK_AUTHORS" ON ("ANNOTATIONS_AUTHOR"."ID" = "ANNOTATIONS_BOOK_AUTHORS"."AUTHOR_ID")
    LEFT OUTER JOIN "ANNOTATIONS_BOOK" T4 ON ("ANNOTATIONS_BOOK_AUTHORS"."BOOK_ID" = T4."ID")
WHERE "ANNOTATIONS_AUTHOR"."AGE" < 30
GROUP BY
    "ANNOTATIONS_AUTHOR"."ID",
    "ANNOTATIONS_AUTHOR"."NAME",
    "ANNOTATIONS_AUTHOR"."AGE",
    CASE WHEN "ANNOTATIONS_BOOK"."ID" IS NULL THEN 1 ELSE 0 END
```
previously
```sql
SELECT
    "ANNOTATIONS_AUTHOR"."NAME",
    CASE WHEN "ANNOTATIONS_BOOK"."ID" IS NULL THEN 0 ELSE MAX(T4."PAGES") END AS "MAX_PAGES"
FROM
    "ANNOTATIONS_AUTHOR"
    LEFT OUTER JOIN "ANNOTATIONS_BOOK" ON ("ANNOTATIONS_AUTHOR"."ID" = "ANNOTATIONS_BOOK"."CONTACT_ID")
    LEFT OUTER JOIN "ANNOTATIONS_BOOK_AUTHORS" ON ("ANNOTATIONS_AUTHOR"."ID" = "ANNOTATIONS_BOOK_AUTHORS"."AUTHOR_ID")
    LEFT OUTER JOIN "ANNOTATIONS_BOOK" T4 ON ("ANNOTATIONS_BOOK_AUTHORS"."BOOK_ID" = T4."ID")
WHERE "ANNOTATIONS_AUTHOR"."AGE" < 30
GROUP BY
    "ANNOTATIONS_AUTHOR"."ID",
    "ANNOTATIONS_AUTHOR"."NAME",
    "ANNOTATIONS_AUTHOR"."AGE",
    "ANNOTATIONS_BOOK"."ID"
```

**Comment by felixxm** [comment]
@Ian-Foote Do you have time to keep working on this? If not I can try to fix remaining comments.

**Comment by Ian-Foote** [comment]
I had a look last night, but I didn't get anywhere (not having a working oracle setup was the biggest blocker).

**Comment by felixxm** [comment]
> I had a look last night, but I didn't get anywhere (not having a working oracle setup was the biggest blocker).

You could check other comments and leave an Oracle issue to me :detective:

**Comment by Ian-Foote** [comment]
Yeah, I'll do that now.

**Comment by felixxm** [comment]
> Yeah, I'll do that now.

Great, thanks :heart:

**Comment by Ian-Foote** [comment]
@felixxm All yours ;)

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@Ian-Foote I fixed Oracle failure by adding `get_group_by_cols()` :relaxed:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@Ian-Foote Thanks for updates :+1: We are almost there :runner: 

I rebased, squashed commits, fix tests failures on Oracle, added new tests `test_lookup_in_order_by`, `test_conditional_expression`, `test_filter_lookup_rhs` and fixed related failures on Oracle, and pushed small edits.

Unfortunately `test_filter_lookup_rhs` crashes on PostgreSQL, it's an issue similar to the 170b006ce82b0ecf26dc088f832538b747ca0115. It seems that LHS should be wrapped in parentheses to respect operator precedence :thinking: (\cc @charettes).

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
> Unfortunately `test_filter_lookup_lhs` crashes on PostgreSQL sob , it's an issue similar to the [170b006](https://github.com/django/django/commit/170b006ce82b0ecf26dc088f832538b747ca0115). It seems that LHS should be wrapped in parentheses to respect operator precedence thinking (\cc @charettes).


Fixed :partying_face:

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/models/lookups.py [hunk 16]
```diff
@@ -72,20 +72,32 @@ def set_source_expressions(self, new_exprs):
...
 
     def process_lhs(self, compiler, connection, lhs=None):
         lhs = lhs or self.lhs
         if hasattr(lhs, 'resolve_expression'):
             lhs = lhs.resolve_expression(compiler.query)
-        return compiler.compile(lhs)
+        sql, params = compiler.compile(lhs)
+        if isinstance(lhs, Lookup):
+            # Wrapped in parentheses to respect operator precedence.
+            sql = f'({sql})'
```
Comment (charettes) [inline django/db/models/lookups.py:99]: I wonder if a crash similar to 170b006 can happen if the lhs is a `Subquery` annotation.

Comment (felixxm) [reply django/db/models/lookups.py:99]: Challenge accepted :wink: I added two more tests `test_filter_exists_lhs` and `test_filter_subquery_lhs`.

**Comment by Ian-Foote** [comment]
Thanks for getting this over the line @felixxm and thanks @charettes, @hannseman and @pauloxnet for the reviews!
